lib/ukstore/store.c

```diff
@@ -29,3 +29,1186 @@
+		if (ret < 0)
+			return ret;
+
+		sscanf_ret = sscanf(val, "%p", (void **) out);
+		if (sscanf_ret < 0)
+			return sscanf_ret;
+
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+}
+
+int
+_uk_store_get_charp(const struct uk_store_entry *e, char **out)
+{
+	int ret;
+	char *str;
+
+	UK_ASSERT(e);
+
+	switch (e->type) {
+	case uk_store_get_entry_type(u8): {
+		__u8 val;
+
+		str = calloc(UK_STORE_NR_SIZE8_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.u8(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		sprintf(str, "%" __PRIu8, val);
+		*out  = str;
+		return ret;
+	}
+
+	case uk_store_get_entry_type(s8): {
+		__s8 val;
+
+		str = calloc(UK_STORE_NR_SIZE8_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.s8(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		sprintf(str, "%" __PRIs8, val);
+		*out  = str;
+		return ret;
+	}
+
+	case uk_store_get_entry_type(u16): {
+		__u16 val;
+
+		str = calloc(UK_STORE_NR_SIZE16_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.u16(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		sprintf(str, "%" __PRIu16, val);
+		*out  = str;
+		return ret;
+	}
+
+	case uk_store_get_entry_type(s16): {
+		__s16 val;
+
+		str = calloc(UK_STORE_NR_SIZE16_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.s16(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		sprintf(str, "%" __PRIs16, val);
+		*out  = str;
+		return ret;
+	}
+
+	case uk_store_get_entry_type(u32): {
+		__u32 val;
+
+		str = calloc(UK_STORE_NR_SIZE32_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.u32(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		sprintf(str, "%" __PRIu32, val);
+		*out  = str;
+		return ret;
+	}
+
+	case uk_store_get_entry_type(s32): {
+		__s32 val;
+
+		str = calloc(UK_STORE_NR_SIZE32_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.s32(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		sprintf(str, "%" __PRIs32, val);
+		*out  = str;
+		return ret;
+	}
+
+	case uk_store_get_entry_type(u64): {
+		__u64 val;
+
+		str = calloc(UK_STORE_NR_SIZE64_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.u64(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		sprintf(str, "%" __PRIu64, val);
+		*out  = str;
+		return ret;
+	}
+
+	case uk_store_get_entry_type(s64): {
+		__s64 val;
+
+		str = calloc(UK_STORE_NR_SIZE64_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.s64(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		sprintf(str, "%" __PRIs64, val);
+		*out  = str;
+		return ret;
+	}
+
+	case uk_store_get_entry_type(uptr): {
+		__uptr val;
+
+		str = calloc(UK_STORE_NR_SIZE64_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.uptr(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		sprintf(str, "%p", (void *) val);
```

Comment:
I would suggest to use `"0x"__PRIx64` as format. `%p` is sometimes differently implementented across libcs. I have seen it once without leading `0x` prefix and once with.

Reply:
I'll just do it like this for that case

```c
		if (strncmp(val, "0x", 2) == 0)
			ret = sscanf(val, "0x%" __SCNx64, (void **) &to_set);
		else
			ret = sscanf(val, "%" __SCNx64, (void **) &to_set);
```